### PR TITLE
Use swap fidelity instead of gate time as a distance metric

### PR DESCRIPTION
**Pull request by karalekas**
And benchmark the peformance.

Replaces stale PR #331

–––––––

Some things to inspect before merge, inherited from #449:

* `chip-specification` sometimes carries extra data used by a particular addressing method to make more intelligent decisions, like the `time-bound` key inside of various `hardware-object`'s `misc-data` slots. It would probably be better to make this part of (the subclasses of) `addresser-state`, since which addresser cares about what is really specific to the style of addressing being used. (But it's also conceivable that the compressor will want to see that same information and optimize along the same `cost-function`...?) (Also, I don't think the `cost-bound` field gets populated at all? correctly? for the fidelity addresser!)
* I did a bunch of monkey-patching around Corwin's contributions to the addresser (viz., the various bits of SWAP-searching), which involved passing different "versions" of the parent `cost-function` around with different argument specializations and so on. It's gross, and it sort-of breaks the abstraction of addresser-state.
* We should very thoroughly test the case of running programs near the saturation limit on _small_ chips, as this is the most common use case today.
* Seriously consider adding terms to the fidelity cost function which represent (T1-type) decay.

### CODE REVIEW

File: benchmarking/rewiring-analysis.lisp [hunk 1]
```diff
@@ -296,3 +297,14 @@
     :a*-depth  (:partial :a*           :a*           1d0 :depth)
     :a*-sum    (:partial :a*           :a*           1d0 :sum)
     ))
+
+(defvar *cost-fn-weight-style-assn*
+    (make-assignments
+        ((*random-state* #+sbcl (sb-ext:seed-random-state 1)
```
Comment (stylewarning) [inline benchmarking/rewiring-analysis.lisp]: i'd prefer to move these into separate function so we know it needs porting

File: benchmarking/rewiring-analysis.lisp [hunk 2]
```diff
@@ -296,3 +297,14 @@
...
+                         #+ecl  (make-random-state 1)
+                         #-(or sbcl ecl) (error "don't know how to seed random state"))
+         (quil::*compressor-passes* 1))
+        (quil::*cost-fn-weight-style*)
+      :duration   (:duration)
+      :fidelity   (:fidelity)
```
Comment (stylewarning) [inline benchmarking/rewiring-analysis.lisp]: close parens

File: src/addresser/logical-schedule.lisp [hunk 3]
```diff
@@ -634,7 +635,7 @@ mapping instructions to their tags. "
     (multiple-value-bind (max-value value-hash)
         (lscheduler-walk-graph lschedule
                                :base-value 1d0
-                               :bump-value #'fidelity-bumper
+                               :bump-value (lambda (instr value) (let ((fid (fidelity-bumper instr value))) (format t "~/cl-quil:instruction-fmt/ -> ~a~%" instr fid) fid))
```
Comment (stylewarning) [inline src/addresser/logical-schedule.lisp]: did you intend to keep the print there

File: src/addresser/cost-function.lisp [hunk 4]
```diff
@@ -40,9 +53,9 @@
     (dotimes (link-index (length (vnth 1 (chip-specification-objects chip-spec))))
       (let ( ; eventually this should look up "SWAP" in this list, but for now
              ; this is guaranteed to be the only 2Q permutation anyway.
-            (weight (permutation-record-duration
-                     (vnth 0 (hardware-object-permutation-gates
-                              (chip-spec-nth-link chip-spec link-index)))))
+            (weight (ecase *cost-fn-weight-style*
+                      (:duration (permutation-record-duration (vnth 0 (hardware-object-permutation-gates (chip-spec-nth-link chip-spec link-index)))))
+                      (:fidelity (- (log (swap-fidelity chip-spec link-index))))))
```
Comment (stylewarning) [inline src/addresser/cost-function.lisp]: can we document this formula for the future reader

**Comment by karalekas** [comment]
Just pushed additional work-in-progress work, still need to address the partial review comments above.

File: benchmarking/rewiring-analysis.lisp [hunk 5]
```diff
@@ -63,7 +63,7 @@
 (defparameter *rewiring-explicit-test-chips*
   '((:0006q-2ring *chip-6q-2ring*)
     (:0008q *chip-0008q*)
-    (:0016q *chip-0016q*)
+    #+ignore(:0016q *chip-0016q*)
```
Comment (stylewarning) [inline benchmarking/rewiring-analysis.lisp]: this shouldn't make it into the final merge

File: src/addresser/cost-function.lisp [hunk 6]
```diff
@@ -55,7 +55,7 @@
              ; this is guaranteed to be the only 2Q permutation anyway.
             (weight (ecase *cost-fn-weight-style*
                       (:duration (permutation-record-duration (vnth 0 (hardware-object-permutation-gates (chip-spec-nth-link chip-spec link-index)))))
-                      (:fidelity (- (log (swap-fidelity chip-spec link-index))))))
+                      (:fidelity (* (permutation-record-duration (vnth 0 (hardware-object-permutation-gates (chip-spec-nth-link chip-spec link-index)))) (- (log (swap-fidelity chip-spec link-index)))))))
```
Comment (stylewarning) [inline src/addresser/cost-function.lisp]: put each factor on a newline

File: src/addresser/logical-schedule.lisp [hunk 7]
```diff
@@ -599,6 +599,7 @@ mapping instructions to their tags. "
      (hash-table-count (lscheduler-earlier-instrs lschedule))))
 
 (defun lscheduler-calculate-fidelity (lschedule chip-spec)
+  (declare (optimize (debug 3)))
```
[the remark this answers is not in the document]
Comment (stylewarning) [reply src/addresser/logical-schedule.lisp]: delete unless needed

File: src/chip-specification.lisp [hunk 8]
```diff
@@ -293,7 +293,7 @@ used to specify CHIP-SPEC."
                                   (hardware-object-gate-information obj))
                          (make-gate-record :duration 150
                                            :fidelity fidelity)))))
-      (dolist (data `((:cz     "CZ"     ()  (_ _) 0.89d0)
+      (dolist (data `((:cz     "CZ"     ()  (_ _) 0.98d0)
```
Comment (stylewarning) [inline src/chip-specification.lisp]: can we write a comment where these numbers come from, or why their choices are reasonable?

File: src/chip-specification.lisp [hunk 9]
```diff
@@ -346,10 +346,10 @@ used to specify CHIP-SPEC."
       (when (member ':RZ type)
         (stash-gate-record "RZ" '(_) (list q) 1/100 +near-perfect-fidelity+))
       (when (member ':X/2 type)
-        (stash-gate-record "RX" '(#.(/ pi 2))  `(,q) 9 0.98d0)
```
Comment (stylewarning) [inline src/chip-specification.lisp]: likewise here

File: src/addresser/fidelity-addresser.lisp [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+;;; itself as a heuristic for producing "efficient" programs, where "efficient"
+;;; is primarily taken to mean "short". In this routine, we make the necessary
+;;; modifications to take "efficient" to mean "high-fidelity".
+;;;
+;;; See DO-FIDELITY-ADDRESSING below for the main entry point.
+
+(defclass fidelity-addresser-state (addresser-state)
+  ())
+
+(defstruct fidelity-cost
```
Comment (stylewarning) [inline src/addresser/fidelity-addresser.lisp]:
```
(defstruct (fidelity-cost (:constructor fidelity-cost (value)))
  (value nil :type real))
```
add `:read-only t` if it's read-only too.

This will change the constructor to just be `(fidelity-cost 10)`

File: src/addresser/fidelity-addresser.lisp [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+        (gate-weights-cost 0d0))
+    
+    ;; calculate log-infidelity coming from INSTR, using recombination:
+    ;;   + calculate the cut point
+    ;;   + calculate the infidelity of instructions since the cut point
+    ;;   + use only the additional infidelity penalty
+    (when (and instr (typep instr 'gate-application))
+      (let* ((chip-spec (addresser-state-chip-specification state))
+             (instruction-expansion
+               (let ((*compress-carefully* nil))
```
Comment (stylewarning) [inline src/addresser/fidelity-addresser.lisp]: why?

File: src/addresser/fidelity-addresser.lisp [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+                 (expand-to-native-instructions (list instr) chip-spec))))
+        (setf instr-cost (- (log (calculate-instructions-fidelity instruction-expansion chip-spec))))
+        
+        ;; then, see if there's a non-naive cost available
+        (a:when-let* ((hardware-object (lookup-hardware-object chip-spec instr))
```
Comment (stylewarning) [inline src/addresser/fidelity-addresser.lisp]: are you sure _all_ of these should participate in `when`? It sounds like it should be
```
(let* (...)
  ;; assert some things are non-null
  (when <actual thing that can be null>
    ...))
```

File: src/addresser/addresser-common.lisp [hunk 13]
```diff
@@ -0,0 +1,788 @@
...
+   "Select a permutation and schedule it for execution. The permutation is selected to lower the
+cost-function associated to the current lschedule.")
+  (:method (state rewirings-tried)
+    (with-slots (lschedule initial-l2p working-l2p chip-sched chip-spec qq-distances)
+        state
+      (format *compiler-noise-stream*
+              "SELECT-AND-EMBED-A-PERMUTATION: entering SWAP selection phase.~%")
+      (let ((gates-in-waiting (assign-weights-to-gates state)))
+        (ecase *addresser-swap-search-type*
+          #+ignore  ; TODO: test this
```
Comment (stylewarning) [inline src/addresser/addresser-common.lisp]: TODO

File: src/addresser/fidelity-addresser.lisp [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+  (double= (fidelity-cost-value val1) (fidelity-cost-value val2)))
+
+(defmethod build-worst-cost ((state fidelity-addresser-state))
+  (make-fidelity-cost :value most-positive-fixnum))
+
+(defmethod assign-weights-to-gates ((state fidelity-addresser-state) &optional (1q-descaling 1/10))
+  (multiple-value-bind (max-value value-hash)
+      (lscheduler-walk-graph (addresser-state-logical-schedule state)
+                             :base-value 0
+                             :bump-value (lambda (instr value)
```
Comment (stylewarning) [inline src/addresser/fidelity-addresser.lisp]: abstract the lambda into an flet

File: src/addresser/fidelity-addresser.lisp [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+                                             (t
+                                              value)))
+                             :test-values #'max)
+    (declare (ignore max-value))
+    value-hash))
+
+(defun initial-fidelity-addresser-working-state (chip-spec initial-rewiring)
+  (let* ((*cost-fn-weight-style* ':fidelity)
+         (state (initial-addresser-working-state chip-spec initial-rewiring)))
+    (change-class state 'fidelity-addresser-state)
```
Comment (stylewarning) [inline src/addresser/fidelity-addresser.lisp]: colour me dubious

File: src/addresser/fidelity-addresser.lisp [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+    state))
+
+(defun do-greedy-fidelity-addressing (instrs
+                               chip-spec
```
Comment (stylewarning) [inline src/addresser/fidelity-addresser.lisp]: indentation

File: src/addresser/outgoing-schedule.lisp [hunk 17]
```diff
@@ -104,6 +104,12 @@ BEFORE-INST to make use of RESOURCE."
                  (not (resource-subsetp (instruction-resources instr) resource)))
         :maximize (chip-schedule-end-time schedule instr)))
 
+(defun chip-schedule-from-carving-point (schedule resource point)
+  (loop :for instr :being :the :hash-keys :of (chip-schedule-times schedule)
```
Comment (stylewarning) [inline src/addresser/outgoing-schedule.lisp]: document

File: src/addresser/temporal-addresser.lisp [hunk 18]
```diff
@@ -4,107 +4,173 @@
...
+(defclass temporal-addresser-state (addresser-state)
+  ((1q-queues :accessor temporal-addresser-state-1q-queues
+              :initarg :1q-queues
+              :documentation "The family of queues where not-yet-scheduled 1Q instructions live, while we get them out of the way to process 2Q instructions.")))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;; Cost function ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defstruct temporal-cost
```
Comment (stylewarning) [inline src/addresser/temporal-addresser.lisp]: document, provide read-only and types qw n33e3e

File: src/addresser/temporal-addresser.lisp [hunk 19]
```diff
@@ -4,107 +4,173 @@
...
-    (if (= q0 qubit) q1 q0)))
-
-;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;; Chip Schedule Management ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; This file contains the subroutines specific to the operation of the version
+;;; of the addresser that uses overall program runtime to optimize output.
+
+(defclass temporal-addresser-state (addresser-state)
+  ((1q-queues :accessor temporal-addresser-state-1q-queues
+              :initarg :1q-queues
+              :documentation "The family of queues where not-yet-scheduled 1Q instructions live, while we get them out of the way to process 2Q instructions.")))
```
Comment (stylewarning) [inline src/addresser/temporal-addresser.lisp]: document class

File: src/addresser/temporal-addresser.lisp [hunk 20]
```diff
@@ -4,107 +4,173 @@
...
+;;;; CZ 3 7
+;;;; on a chip with a linear circuit topology, and (for whatever reason) assuming
+;;;; qubits 0 and 7 to be fixed, a naive sum of the SWAP distances will result
+;;;; in a standstill. the 0-3 distance is 2 SWAPs and the 3-7 distance is 3 SWAPs,
+;;;; and trading 3 for 2 or 3 for 4 preserves the total number of 1+4 = 3+2 = 5
+;;;; SWAPs in all. to break this kind of deadlock, we make very distant gates
+;;;; suffer a decay penalty, so that gates that are already nearer are preferred
+;;;; and executed first.
+
+(defparameter *cost-fn-tier-decay* 0.5
```
Comment (stylewarning) [inline src/addresser/temporal-addresser.lisp]: defvar for configuration (if it's so)

File: src/addresser/temporal-addresser.lisp [hunk 21]
```diff
@@ -523,550 +302,118 @@ cost-function associated to the current lschedule."
...
+     ;; dequeue and set the dirty bit  
+     (lscheduler-dequeue-instruction (addresser-state-logical-schedule state) instr)
+     (push instr (aref (temporal-addresser-state-1q-queues state)
+                       (qubit-index (first (application-arguments instr)))))
+     (values t nil nil))
+    (t
+     (call-next-method))))
+
+;; flush the queues before adding the 2Q instruction
+(defmethod append-instr-to-chip-schedule :before
```
Comment (stylewarning) [inline src/addresser/temporal-addresser.lisp]: why do we want to do this in a before method?

File: src/compilation-methods.lisp [hunk 22]
```diff
@@ -248,8 +248,9 @@ Returns a value list: (processed-program, of type parsed-program
 
          (touch-unpreserved-block (blk registrant)
            ;; actually process this block
-           (multiple-value-bind (initial-l2p chip-schedule final-l2p)
-               (do-greedy-temporal-addressing
+           (multiple-value-bind (chip-schedule initial-l2p final-l2p)
+               ;; TODO: here, and below, we need to find a way to not hard-code this method call
```
Comment (stylewarning) [inline src/compilation-methods.lisp]: what does this mean?

File: tests/compiler-hook-tests.lisp [hunk 23]
```diff
@@ -477,7 +477,7 @@ MEASURE 1
 (deftest test-clever-CCNOT-depth-reduction ()
   "Test that the ':GREEDY-QUBIT swap selection strategy brings CZ depth down to optimal for CCNOT."
   (let ((p (quil::compiler-hook (quil::parse-quil "
-PRAGMA INITIAL_REWIRING \"GREEDY\"
```
Comment (stylewarning) [inline tests/compiler-hook-tests.lisp]: why?

File: tests/initial-rewiring-tests.lisp [hunk 24]
```diff
@@ -117,6 +117,7 @@ HALT
     ;; It's questionable whether asserting an error is the right thing to do here, but it at least
     ;; documents the current behavior. And if we ever improve NAIVE rewiring so that this doesn't
     ;; error, this test will fail and remind us to update it to NOT-SIGNALS.
+    #+ignore
```
Comment (stylewarning) [inline tests/initial-rewiring-tests.lisp]: oof

File: src/addresser/addresser-common.lisp [hunk 25]
```diff
@@ -0,0 +1,788 @@
...
+;;; The potential heuristics used to search for solutions to the addressing
+;;; problem are manifold, and different ones are called for in different
+;;; situations. This file contains the guts common to each such method, and
+;;; different methods will invoke these guts after binding particular behaviors
+;;; to various hooks.
+;;;
+;;; See DO-GREEDY-TEMPORAL-ADDRESSING below for the main entry point.
+
+
+(defparameter *addresser-swap-lookahead-depth* 2
```
Comment (stylewarning) [inline src/addresser/addresser-common.lisp]: defvar for config vars

File: src/addresser/addresser-common.lisp [hunk 26]
```diff
@@ -0,0 +1,788 @@
...
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;; Addresser State Updates ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+;;; The main entry point for temporal addressing is DO-GREEDY-TEMPORAL-ADDRESSING
+;;; below. However, this manages a fair amount of state as it navigates the
+;;; addressing process. The struct below bundles this together for the sake of
+;;; convenience. At any given point of the execution there is only one of these
+;;; objects, with values being mutated along the way.
+
+(defclass addresser-state ()
```
Comment (stylewarning) [inline src/addresser/addresser-common.lisp]: can we add types to the slots, or at least document them in the strings?

File: src/addresser/addresser-common.lisp [hunk 27]
```diff
@@ -0,0 +1,788 @@
...
+  (let* ((hardware-object (chip-spec-nth-link chip-spec link-index))
+         (permutation-record (vnth 0 (hardware-object-permutation-gates
+                                      hardware-object)))
+         (swap (apply #'build-gate
+                      (permutation-record-operator permutation-record)
+                      '()
+                      (coerce (chip-spec-qubits-on-link chip-spec link-index) 'list))))
+    (calculate-instructions-fidelity (expand-to-native-instructions (list swap) chip-spec) chip-spec)))
+
+(defparameter *cost-fn-weight-style* ':duration)
```
Comment (stylewarning) [inline src/addresser/addresser-common.lisp]: defvar

File: src/addresser/addresser-common.lisp [hunk 28]
```diff
@@ -0,0 +1,788 @@
...
+                   :working-l2p (copy-rewiring initial-l2p)
+                   :lschedule (make-lscheduler)
+                   :qq-distances (precompute-qubit-qubit-distances chip-spec)
+                   :qubit-cc (a:extremum (chip-spec-live-qubit-cc chip-spec)
+                                         #'>
+                                         :key #'length)
+                   :chip-sched (make-chip-schedule chip-spec)
+                   :chip-spec chip-spec)))
+
+;; TODO: why isn't this part of the state?
```
Comment (stylewarning) [inline src/addresser/addresser-common.lisp]: defvar

File: src/addresser/addresser-common.lisp [hunk 29]
```diff
@@ -0,0 +1,788 @@
...
+(defparameter *addresser-use-free-swaps* nil
+  "Does the addresser treat the initial rewiring as something that can be changed?")
+
+(defgeneric assign-weights-to-gates (state)
+  (:documentation "Generic method for assigning weights to gates, for consumption by COST-FUNCTION."))
+
+(defgeneric cost-function (state &key gate-weights instr)
```
Comment (stylewarning) [inline src/addresser/addresser-common.lisp]: it seems this sometimes returns NIL. Is that true?

Should these be keyword arguments or should they be real args so we can dispatch on `instr`?

This function is super odd.

File: src/addresser/addresser-common.lisp [hunk 30]
```diff
@@ -0,0 +1,788 @@
...
+                                     :instr instr))))
+             (push (copy-rewiring working-l2p) rewirings-tried)
+             (embed-swap (select-cost-lowering-swap working-l2p chip-spec #'cost-function rewirings-tried)
+                         initial-l2p
+                         working-l2p
+                         chip-spec
+                         chip-sched
+                         :use-free-swaps *addresser-use-free-swaps*)
+             rewirings-tried))
+          ;; TODO: reenable this
```
Comment (stylewarning) [inline src/addresser/addresser-common.lisp]: TODO

File: benchmarking/rewiring-analysis.lisp [hunk 31]
```diff
@@ -296,3 +297,14 @@
...
+(defvar *cost-fn-weight-style-assn*
+    (make-assignments
+        ((*random-state* #+sbcl (sb-ext:seed-random-state 1)
+                         #+ecl  (make-random-state 1)
+                         #-(or sbcl ecl) (error "don't know how to seed random state"))
+         (quil::*compressor-passes* 1))
+        (quil::*cost-fn-weight-style*)
+      :duration   (:duration)
+      :fidelity   (:fidelity)
+      ))
```
[the remark this answers is not in the document]
Comment (ecpeterson) [reply benchmarking/rewiring-analysis.lisp]: when in rome

File: src/addresser/fidelity-addresser.lisp (hunk 11 above)
Comment (ecpeterson) [reply src/addresser/fidelity-addresser.lisp]: I _think_ (but I also forget, since I failed to leave an explanatory comment!) that `*compress-carefully*` interacts poorly with `application-thread-invocation`s, since those expand into instruction sequences but don't themselves have associated matrices to check against.

File: src/addresser/fidelity-addresser.lisp (hunk 12 above)
Comment (ecpeterson) [reply src/addresser/fidelity-addresser.lisp]: Definitely this is a case of not wanting to push the indentation level further and further up. The first two really belong to the `when`, the rest to a `let*`. I'll break it apart.

File: src/addresser/fidelity-addresser.lisp (hunk 15 above)
Comment (ecpeterson) [reply src/addresser/fidelity-addresser.lisp]: I'm almost definitely abusing CLOS here. What I want is a (fairly flat) class tree w/ a corresponding tree of custom constructors which all behave like `:after` specializations. This is one way to achieve that, but it's probably not the right one. What's the right one?

File: src/addresser/temporal-addresser.lisp (hunk 18 above)
Comment (ecpeterson) [reply src/addresser/temporal-addresser.lisp]: idk "qw n33e3e"

File: src/addresser/temporal-addresser.lisp (hunk 20 above)
Comment (ecpeterson) [reply src/addresser/temporal-addresser.lisp]: Not quite: it's a tunable parameter, and tuning it can get the addresser out of deadlocks, but it's very unclear _how_ to successfully tune it, and so we just randomize it across different addresser loop iterations to keep out of trouble.

File: src/addresser/temporal-addresser.lisp (hunk 21 above)
Comment (ecpeterson) [reply src/addresser/temporal-addresser.lisp]: The `1q-queues` are populated with single-qubit instructions that we deferred scheduling so that we could inspect 2Q instructions further down the line. If we're calling `append-instr-to-chip-schedule`, then we're actually trying to schedule a (2Q) instruction, which means we can no longer defer any intersecting 1Q instructions—which have to be handled `:before` we handle the requested 2Q instruction.

File: src/addresser/addresser-common.lisp (hunk 29 above)
Comment (ecpeterson) [reply src/addresser/addresser-common.lisp]: I'm definitely open to reconsidering this function, but I don't think it's possible for it to return NIL, and I do think the docstring makes that clear.

File: src/addresser/fidelity-addresser.lisp (hunk 15 above)
Comment (stylewarning) [reply src/addresser/fidelity-addresser.lisp]: Why not define `after` methods on all of the classes?

File: src/chip-specification.lisp [hunk 32]
```diff
@@ -369,6 +360,8 @@ used to specify CHIP-SPEC."
           (declare (ignore chip-spec))
           (when (optimal-2q-target-meets-requirements arch ':cz)
             'cnot-to-native-cnots))
+        (constantly 'cz-to-native-czs)
+        #+ignore
```
Comment (stylewarning) [inline src/chip-specification.lisp]: why is this ignored

File: tests/compilation-tests.lisp [hunk 33]
```diff
@@ -165,4 +165,7 @@ CNOT 0 2"))
     (is (every (link-nativep chip) 2q-code))
     ;; NOTE: Decomposing into fewer 2q gates is more of a regression
     ;; test on quality of compilation, and not on correctness.
-    (is (>= 6 (length 2q-code)))))
+
+    ;; NOTE: This test used to be upper-bounded 6. Changes to the
```
Comment (stylewarning) [inline tests/compilation-tests.lisp]: sounds like an `XXX TODO:` not a note; maybe it should warn if it's not < 7?

Comment (ecpeterson) [reply tests/compilation-tests.lisp]: seconded

File: tests/initial-rewiring-tests.lisp (hunk 24 above)
Comment (ecpeterson) [reply tests/initial-rewiring-tests.lisp]: seconded

Comment (ecpeterson) [reply tests/initial-rewiring-tests.lisp]: I updated this to `not-signals` instead, but that only makes me think that the bad test case has changed a bit & so this test isn't testing anything anymore. 🤷‍♂

Comment (appleby) [reply tests/initial-rewiring-tests.lisp]: fwiw I'm fine with removing it if it's not a useful test

Comment (ecpeterson) [reply tests/initial-rewiring-tests.lisp]: I don't think that the test is bad—it captures a real failure of the naive addresser!—but it's fragile, in the sense that changes to the addresser stand a high chance of changing where the edge cases are, and this test hard-codes an edge case in.

Comment (appleby) [reply tests/initial-rewiring-tests.lisp]: I'd vote to remove then. imo fragile test == bad test :-)

Comment (appleby) [reply tests/initial-rewiring-tests.lisp]: I had a feeling it might be fragile when I added it, but wasn't sure... so I won't shed any tears to see it go!

File: benchmarking/qpu-rewiring/Aspen-32Q-Y.qpu [hunk 34]
```diff
@@ -0,0 +1,509 @@
+{
+  "name": "Aspen-4-16Q-A",
```
[the remark this answers is not in the document]
Comment (ecpeterson) [reply benchmarking/qpu-rewiring/Aspen-32Q-Y.qpu]: it won't hurt to update it, but the `name` field is mostly meant for over-the-wire ISA payloads, so it's allowed to be independent of the file name (which might not exist at all)

File: src/compilation-methods.lisp [hunk 35]
```diff
@@ -147,6 +147,8 @@
 ;; Forward declaration from compressor.lisp
 (declaim (special *compressor-passes*))
 
+(defparameter *addresser-state-constructor* 'initial-fidelity-addresser-working-state)
```
[the remark this answers is not in the document]
Comment (ecpeterson) [reply src/compilation-methods.lisp]: I actually am inclined to make this the default for now, though it would be smart to expose some kind of toggle through the application layer. I'm inclined to promote that to a fresh Issue rather than deal with it now.

File: src/addresser/fidelity-addresser.lisp (hunk 15 above)
Comment (ecpeterson) [reply src/addresser/fidelity-addresser.lisp]: OK, I feel good about this now. Let me know what you think.

**Comment by ecpeterson** [comment]
Reminder that much of this PR is refactor smoke and mirrors: `embed-swap.lisp`, `addresser-common.lisp`, and `temporal-addresser.lisp` are 99% copy/pasted from the old `temporal-addresser.lisp`, with some `defun`s changed to `defmethod`s. Don't be put off by the 3.7k line count.

File: src/addresser/path-heuristic.lisp [hunk 36]
```diff
@@ -77,35 +77,34 @@ rewiring."
...
+    :for gate :being :the :hash-keys :of gates-in-waiting
+    :for tier := (gethash gate gates-in-waiting)
+    :for tier-weight := (expt 0.5d0 tier)
 
-    :for tier :in gates-in-waiting
-    :and tier-index :below 3
-    :and tier-weight := 1d0 :then (* tier-weight 0.5d0)
+    :do (when (and (typep gate 'gate-application)
+                   (= 2 (length (application-arguments gate))))
+          (destructuring-bind (q0 q1) (mapcar #'qubit-index (application-arguments gate))
```
Comment (appleby) [inline src/addresser/path-heuristic.lisp]: another chance for `application-qubit-indices` to shine!

honestly, `(mapcar #'quibit-index ...)` is pretty common in quilc sources (probably predating the above function), and not that much longer / harder to read. feel free to leave it.

there is even a duplicate definition called `application-argument-indicies` in `compilation-tests.lisp`  :-)

**Comment by notmgsk** [comment]
> Reminder that much of this PR is refactor smoke and mirrors: `embed-swap.lisp`, `addresser-common.lisp`, and `temporal-addresser.lisp` are 99% copy/pasted from the old `temporal-addresser.lisp`, with some `defun`s changed to `defmethod`s. Don't be put off by the 3.7k line count.

Figuring out where that 1% lies is a PITA.

**Comment by ecpeterson** [comment]
I agree. If we’d had the foresight to do this refactor before doing the
fidelity stuff, we surely would have. Alas: we are merely programmers.

**Comment by ecpeterson** [comment]
@appleby: Good point / idea. Rather than putting in extra lööps to re-run all the existing tests over both addresser styles, I put in a more meager test which exercises the different styles on a single, simple-enough test input.

File: src/addresser/addresser-common.lisp [hunk 37]
```diff
@@ -0,0 +1,763 @@
...
+        ;; for each right vertex...
+        (dotimes (j vertex-count)
+          (let ((ij (aref dist i j))
+                (ik (aref dist i k))
+                (kj (aref dist k j)))
+            ;; is our current best path i->j worse than the concatenation of the
+            ;; current best paths i->k and k->j?
+            (when (> ij (+ ik kj))
+              ;; then replace the old best path with these new ones.
+              (setf (aref dist i j) (+ ik kj))
```
Comment (notmgsk) [inline src/addresser/addresser-common.lisp]: You could totally collapse these `setf`s and impress @jmbr.

File: src/addresser/addresser-common.lisp [hunk 38]
```diff
@@ -0,0 +1,763 @@
...
+         (swap (apply #'build-gate
+                      (permutation-record-operator permutation-record)
+                      '()
+                      (coerce (vnth 0 (hardware-object-cxns hardware-object)) 'list))))
+    (calculate-instructions-fidelity (expand-to-native-instructions (list swap) chip-spec) chip-spec)))
+
+;; nearly ripped straight out of the Wikipedia article for Floyd-Warshall
+(defun precompute-qubit-qubit-distances (chip-spec initial-map)
+  "Implements Floyd-Warshall to compute the minimum weighted distance between any pair of qubits on a CHIP-SPECification, weighted by swap duration."
+  (let* ((vertex-count (length (vnth 0 (chip-specification-objects chip-spec))))
```
Comment (notmgsk) [inline src/addresser/addresser-common.lisp]: `chip-spec-qubits`

File: src/addresser/addresser-common.lisp [hunk 39]
```diff
@@ -0,0 +1,763 @@
...
+(defun next-best-instruction (state instrs)
+  "Find the instruction from the given INSTRS which has the best COST-FUNCTION valuatiuon. Returns three values: the instruction, the index of the hardware object it lies on, and the optional qubit assignments if the selected instruction was only partially assigned."
+  (with-slots (chip-spec chip-sched lschedule) state
+    (let ((best-cost (build-worst-cost state))
+          instr
+          hardware-index)
+      
+      (loop
+        :for (candidate-hardware-index candidate-instr) :in instrs
+        :for physical-qubits := (coerce (vnth 0
```
Comment (notmgsk) [inline src/addresser/addresser-common.lisp]: 🤢

Comment (ecpeterson) [reply src/addresser/addresser-common.lisp]: I agree, but it is what it is.

File: tests/initial-rewiring-tests.lisp [hunk 40]
```diff
@@ -117,7 +117,7 @@ HALT
     ;; It's questionable whether asserting an error is the right thing to do here, but it at least
     ;; documents the current behavior. And if we ever improve NAIVE rewiring so that this doesn't
     ;; error, this test will fail and remind us to update it to NOT-SIGNALS.
```
Comment (stylewarning) [inline tests/initial-rewiring-tests.lisp]: add follow-on comment, but keep the paper trail!

File: src/compilation-methods.lisp [hunk 41]
```diff
@@ -147,6 +147,8 @@
 ;; Forward declaration from compressor.lisp
 (declaim (special *compressor-passes*))
 
+(defparameter *default-addresser-state-class* 'fidelity-addresser-state)
```
Comment (stylewarning) [inline src/compilation-methods.lisp]: probably want to `DEFVAR` since it's configuration

File: src/addresser/temporal-addresser.lisp [hunk 42]
```diff
@@ -523,548 +308,156 @@ cost-function associated to the current lschedule."
...
+                      (setf (gethash hw (temporal-addresser-state-cost-bounds instance))
+                            (temporal-cost-heuristic-value
+                             (cost-function instance :gate-weights (assign-weights-to-gates instance)))))
+        :finally (setf (addresser-state-initial-l2p instance) old-initial-l2p
+                       (addresser-state-working-l2p instance) old-working-l2p
+                       (addresser-state-logical-schedule instance) old-lscheduler
+                       (addresser-state-chip-schedule instance) old-chip-sched)))
+
+;; also, we randomize the cost function weights during select-and-embed-a-permutation
+(defmethod select-and-embed-a-permutation ((state temporal-addresser-state) rewiring-tried)
```
Comment (stylewarning) [inline src/addresser/temporal-addresser.lisp]: it probably makes more sense to make this an `:around` method

File: src/addresser/outgoing-schedule.lisp [hunk 43]
```diff
@@ -104,6 +104,18 @@ BEFORE-INST to make use of RESOURCE."
                  (not (resource-subsetp (instruction-resources instr) resource)))
         :maximize (chip-schedule-end-time schedule instr)))
 
+(defun chip-schedule-from-carving-point (schedule resource point)
+  "Returns the (time-ordered) subsequence of instructions from SCHEDULE which intersect RESOURCE and which start on or after the time POINT.
+
+If POINT is generated by CHIP-SCHEDULE-RESOURCE-CARVING-POINT, then the resulting subsequence has a further \"information light-cone\" property: SCHEDULE will have no other instructions which can receive information from RESOURCE from after POINT."
+  (sort
```
Comment (stylewarning) [inline src/addresser/outgoing-schedule.lisp]: add a comment that this might be optimized by computing the `GETHASH`'s once a la
```
(loop :for instr :being the :hash-keys :of ...
        :using (hash-value v)
  ...
      :collect (cons v instr))
```
then you can sort on `:key #'car` and do a `mapcar #'cdr` on the whole affair.

What is this called again? Schwartzian transform or something? It's some perl hack IIRC.

[ No need to actually optimize it though, but it might be low hanging fruit in the future. ]
